### PR TITLE
MySql receiver POC

### PR DESCRIPTION
**Pull request by JKashyap96**

### CODE REVIEW

**Comment by JKashyap96** [comment]
**MYSQL Reference**:
mysql> select * from Persons where PersonID = 3;
+----------+----------+-----------+----------+----------+
| PersonID | LastName | FirstName | Address  | City     |
+----------+----------+-----------+----------+----------+
|        3 | Kalita   | Raktim    | Hatigaon | Guwahati |
+----------+----------+-----------+----------+----------+
1 row in set (0.01 sec)

**config.yaml**:
extensions:
  sumologic/2:
    collector_name: DB_Collector_TEST_1
    collector_category: collector_category
    install_token: ****************************************
    api_base_url: https://stag-open-events.sumologic.net
exporters:
  logging:
    loglevel: debug
    sampling_initial: 0
    sampling_thereafter: 1
  sumologic:
    auth:
      authenticator: sumologic/2
    source_category: "DB_Collector"
    source_name: "http input"
    source_host: "mysqlRecords"
    compress_encoding: ""
    metadata_attributes:
      - body
    sending_queue:
      enabled: true
    log_format: otlp
receivers:
  mysql:
    endpoint: localhost:3306
    username: Jayanta
    password: ********
    database: otels
    collection_interval: 10s
    query: "Select * from Persons where PersonID = 3;"
service:
  extensions: [sumologic/2]
  pipelines:
    logs:
      exporters: [logging,sumologic]
      receivers: [mysql]

**Collector response**:
2022-04-25T12:49:41.181Z	info	sumologicexporter@v0.0.0-00010101000000-000000000000/exporter.go:110	Sumo Logic Exporter configured	{"kind": "exporter", "name": "sumologic", "log_format": "otlp", "metric_format": "otlp", "trace_format": "otlp"}
2022-04-25T12:49:41.181Z	info	builder/exporters_builder.go:255	Exporter was built.	{"kind": "exporter", "name": "sumologic"}
2022-04-25T12:49:41.181Z	info	builder/exporters_builder.go:255	Exporter was built.	{"kind": "exporter", "name": "logging"}
2022-04-25T12:49:41.181Z	info	builder/pipelines_builder.go:224	Pipeline was built.	{"kind": "pipeline", "name": "logs"}
2022-04-25T12:49:41.181Z	info	builder/receivers_builder.go:226	Receiver was built.	{"kind": "receiver", "name": "mysql", "datatype": "logs"}
2022-04-25T12:49:41.181Z	info	service/service.go:82	Starting extensions...
2022-04-25T12:49:41.181Z	info	extensions/extensions.go:38	Extension is starting...	{"kind": "extension", "name": "sumologic/2"}
2022-04-25T12:49:41.181Z	info	credentials/credentialsstore_localfs.go:146	Collector registration credentials retrieved from local fs	{"kind": "extension", "name": "sumologic/2", "path": "/home/vagrant/.sumologic-otel-collector/e5b43459fb077ecd26fe90f46c820df78a4217685296583725042406460fa8e7"}
2022-04-25T12:49:41.181Z	info	sumologicextension@v0.0.54-beta.0/extension.go:188	Validating collector credentials...	{"kind": "extension", "name": "sumologic/2", "collector_credential_id": "tHTkNJtFQzbEnHQCSa71", "collector_id": "0000000006645B66"}
2022-04-25T12:49:48.669Z	info	sumologicextension@v0.0.54-beta.0/extension.go:255	Found stored credentials, skipping registration	{"kind": "extension", "name": "sumologic/2", "collector_name": "DB_Collector_TEST_1"}
2022-04-25T12:49:48.669Z	info	extensions/extensions.go:42	Extension started.	{"kind": "extension", "name": "sumologic/2"}
2022-04-25T12:49:48.669Z	info	service/service.go:87	Starting exporters...
2022-04-25T12:49:48.669Z	info	builder/exporters_builder.go:40	Exporter is starting...	{"kind": "exporter", "name": "sumologic"}
2022-04-25T12:49:48.670Z	info	builder/exporters_builder.go:48	Exporter started.	{"kind": "exporter", "name": "sumologic"}
2022-04-25T12:49:48.670Z	info	builder/exporters_builder.go:40	Exporter is starting...	{"kind": "exporter", "name": "logging"}
2022-04-25T12:49:48.670Z	info	builder/exporters_builder.go:48	Exporter started.	{"kind": "exporter", "name": "logging"}
2022-04-25T12:49:48.670Z	info	service/service.go:92	Starting processors...
2022-04-25T12:49:48.670Z	info	builder/pipelines_builder.go:54	Pipeline is starting...	{"kind": "pipeline", "name": "logs"}
2022-04-25T12:49:48.670Z	info	builder/pipelines_builder.go:65	Pipeline is started.	{"kind": "pipeline", "name": "logs"}
2022-04-25T12:49:48.670Z	info	service/service.go:97	Starting receivers...
2022-04-25T12:49:48.670Z	info	builder/receivers_builder.go:68	Receiver is starting...	{"kind": "receiver", "name": "mysql"}
2022-04-25T12:49:48.670Z	info	sumologicextension@v0.0.54-beta.0/extension.go:490	Heartbeat loop initialized. Starting to send hearbeat requests	{"kind": "extension", "name": "sumologic/2", "collector_name": "DB_Collector_TEST_1", "collector_id": "0000000006645B66"}
2022-04-25T12:49:48.672Z	INFO	loggingexporter/logging_exporter.go:69	LogsExporter	{"#logs": 1}
2022-04-25T12:49:48.672Z	DEBUG	loggingexporter/logging_exporter.go:79	ResourceLog #0
Resource SchemaURL:
ScopeLogs #0
ScopeLogs SchemaURL:
InstrumentationScope
LogRecord #0
Timestamp: 1970-01-01 00:00:00 +0000 UTC
Severity:
ShortName:
Body: {"Address":"Hatigaon","City":"Guwahati","FirstName":"Raktim","LastName":"Kalita","PersonID":"3"}
Trace ID:
Span ID:
Flags: 0

2022-04-25T12:49:48.672Z	info	builder/receivers_builder.go:73	Receiver started.	{"kind": "receiver", "name": "mysql"}
2022-04-25T12:49:48.672Z	info	service/telemetry.go:109	Setting up own telemetry...
2022-04-25T12:49:48.672Z	info	service/telemetry.go:129	Serving Prometheus metrics	{"address": ":8888", "level": "basic", "service.instance.id": "27078208-e3bc-4311-8278-c565d0bcc93e", "service.version": "latest"}
2022-04-25T12:49:48.673Z	info	service/collector.go:252	Starting otelcol-sumo...	{"Version": "v0.48.0-sumo-0-7-g00dfa1d9b6", "NumCPU": 8}
2022-04-25T12:49:48.673Z	info	service/collector.go:142	Everything is ready. Begin running and processing data.
^C2022-04-25T12:50:11.777Z	info	service/collector.go:173	Received signal from OS	{"signal": "interrupt"}
2022-04-25T12:50:11.777Z	info	service/collector.go:268	Starting shutdown...
2022-04-25T12:50:11.777Z	info	service/service.go:117	Stopping receivers...
2022-04-25T12:50:11.777Z	info	service/service.go:122	Stopping processors...
2022-04-25T12:50:11.777Z	info	builder/pipelines_builder.go:73	Pipeline is shutting down...	{"kind": "pipeline", "name": "logs"}
2022-04-25T12:50:11.777Z	info	builder/pipelines_builder.go:77	Pipeline is shutdown.	{"kind": "pipeline", "name": "logs"}
2022-04-25T12:50:11.777Z	info	service/service.go:127	Stopping exporters...
2022-04-25T12:50:11.777Z	info	service/service.go:132	Stopping extensions...
2022-04-25T12:50:11.778Z	info	service/collector.go:282	Shutdown complete.

**SUMO JSON response**:
Screenshot 2022-04-25 at 6 30 04 PM

File: pkg/receiver/mysqlreceiver/client.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+func (c *mySQLClient) Connect() error {
+	clientDB, err := sql.Open("mysql", c.connStr)
+	if err != nil {
+		return fmt.Errorf("unable to connect to database: %w", err)
+	}
+	c.client = clientDB
+	return nil
+}
+
+// getInnodbStats queries the db for innodb metrics.
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/client.go]: comment should start with the name of function

File: pkg/receiver/mysqlreceiver/client.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+func (c *mySQLClient) getRecords() (map[int]string,error) {
+	query := "select * from Persons;"
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/client.go]: I understand this only for POC purpose, right?
Please remember to protect your code against SQL injection.

File: pkg/receiver/mysqlreceiver/client.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	return Query(*c, query)
+}
+
+func Query(c mySQLClient, query string) (map[int]string,error) {
+	rows, err := c.client.Query(query)
+	if err != nil {
+		log.Printf("error in executing sql query")
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/client.go]: Please use [go.uber.org/zap](https://pkg.go.dev/go.uber.org/zap) for logging, I see that it is commonly used in OTC, you can find example usage in other receivers

File: pkg/receiver/mysqlreceiver/client.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	}
+	defer rows.Close()
+
+	// Get column names
+	columns, err := rows.Columns()
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/client.go:135]: here `err` is not checked

File: pkg/receiver/mysqlreceiver/client.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	columnstype, err := rows.ColumnTypes()
+
+	if err != nil {
+		log.Printf("error getting column datatype from table , error: %v",  err)
+	} else {
+		var columndatatypes []string
+		for _, colvalue := range columnstype {
+			columndatatypes = append(columndatatypes, colvalue.DatabaseTypeName())
+		}
+		//log.Printf("Column types are: %s", columndatatypes)
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/client.go]: please remove commented code

File: pkg/receiver/mysqlreceiver/client.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		log.Println(err)
+	}	
+	myjsonobject := make(map[string]string)
+	myEntireRecord := make(map[int]string)
+	for j, value := range lines {
+		for i, v := range value {
+			myjsonobject[columns[i]] = v
+		}
+		jsonObjRecord, err := json.Marshal(myjsonobject)
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/client.go:185]: this `err` is not checked

File: pkg/receiver/mysqlreceiver/client.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+type mySQLClient struct {
+	connStr string
+	client  *sql.DB
+}
+
+var _ client = (*mySQLClient)(nil)
+
+func newMySQLClient(conf *Config) client {
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/client.go]: all unexported function should have a comment

File: pkg/receiver/mysqlreceiver/config.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+type Config struct {
+        //scraperhelper.ScraperControllerSettings `mapstructure:",squash"`
+	*config.ReceiverSettings `mapstructure:"-"`
+        Username                                string `mapstructure:"username,omitempty"`
+        Password                                string `mapstructure:"password,omitempty"`
+        Database                                string `mapstructure:"database,omitempty"`
+        AllowNativePasswords                    bool   `mapstructure:"allow_native_passwords,omitempty"`
+        confignet.NetAddr                       `mapstructure:",squash"`
+        CollectionInterval                      string `mapstructure:"collection_interval,omitempty"`
+}
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/config.go]: missing new line

File: pkg/receiver/mysqlreceiver/factory.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
+package mysqlreceiver 
+
+import (
+	"context"
+	//"time"
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/factory.go]: please remove commented code

File: pkg/receiver/mysqlreceiver/receiver.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+	"go.opentelemetry.io/collector/model/pdata"
+)
+
+//var records map[int]string
+
+type mySQLReceiver struct {
+	sqlclient client
+	logger    *zap.Logger
+	config    *Config
+	//data	  map[int]string
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/receiver.go]: please remove commented code

File: pkg/receiver/mysqlreceiver/receiver.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+	consumer consumer.Logs
+}
+
+func newMySQLReceiver (logger *zap.Logger, conf *Config, next consumer.Logs) (component.LogsReceiver, error) {
+
+	return &mySQLReceiver{
+		consumer: next,
+		logger: logger,
+		config: conf,
+		//data: records,
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/receiver.go]: please remove commented code

File: pkg/receiver/mysqlreceiver/receiver.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+		return err
+	}
+	m.sqlclient = sqlclient
+
+	data,err := m.sqlclient.getRecords()
+	if err != nil {
+		m.logger.Error("Failed to fetch records", zap.Error(err))
+		return err
+	}
+	//records = data
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/receiver.go]: please remove commented code

File: pkg/receiver/mysqlreceiver/receiver.go [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+func (m *mySQLReceiver) Start(ctx context.Context, host component.Host) error {
+	sqlclient := newMySQLClient(m.config)
+
+	err := sqlclient.Connect()
+	if err != nil {
+		return err
+	}
+	m.sqlclient = sqlclient
+
+	data,err := m.sqlclient.getRecords()
```
Comment (kasia-kujawa) [inline pkg/receiver/mysqlreceiver/receiver.go]: missing space, between `data` and `err`, please get familiar with linters for go

File: pkg/receiver/mysqlreceiver/client.go (hunk 2 above)
Comment (JKashyap96) [reply pkg/receiver/mysqlreceiver/client.go]: Will be doing that as the project progresses.

File: pkg/receiver/mysqlreceiver/client.go (hunk 6 above)
Comment (JKashyap96) [reply pkg/receiver/mysqlreceiver/client.go:185]: done

**Comment by JKashyap96** [comment]
Proper SUMO JSON response : _code=ZUA6DEfsnv3eRKpsATNOTwbnqf4T01bJyh9FrCtK

File: pkg/receiver/mysqlreceiver/client.go (hunk 2 above)
Comment (JKashyap96) [reply pkg/receiver/mysqlreceiver/client.go]: Have updated the code to read the query from config.yaml for now.

**Comment by JKashyap96** [comment]
Merging for next stage of development.

**Pull request merged**